fuzz_test.go

```diff
@@ -0,0 +1,103 @@
+package smt
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/binary"
+	"math"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func FuzzSMT(f *testing.F) {
+	seeds := [][]byte{
+		[]byte(""),
+		[]byte("foo"),
+		{1, 2, 3, 4},
+		[]byte("\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"),
+		nil,
+	}
+	for _, s := range seeds {
+		f.Add(s)
+	}
+	f.Fuzz(func(t *testing.T, input []byte) {
+		t.Log(input)
+		smn := NewSimpleMap()
+		tree := NewSparseMerkleTree(smn, sha256.New())
+
+		r := bytes.NewReader(input)
+		var keys [][]byte
+		key := func() []byte {
+			if readByte(r) < math.MaxUint8/2 {
+				k := make([]byte, readByte(r)/2)
+				if _, err := r.Read(k); err != nil {
+					return nil
+				}
+				keys = append(keys, k)
+				return k
+			}
+
+			if len(keys) == 0 {
+				return nil
+			}
+
+			return keys[int(readByte(r))%len(keys)]
+		}
+
+		for i := 0; r.Len() != 0; i++ {
+			b, err := r.ReadByte()
+			if err != nil {
+				continue
+			}
+
+			op := op(int(b) % int(Noop))
+			switch op {
+			case Get:
+				_, err := tree.Get(key())
+				if err != nil {
```

Comment:
How do we know when we expect an error and when we don't? Seems like the tests don't assert on "there shouldn't be an error"

Reply:
So when I ported this fuzz over to go native fuzzing the ops will all "fail" as they aren't being used in the way they would in production. The way the fuzzes were written was (I assume) to just try and break not make sure the features return the correct values etc. This will be iterated upon later as I work on the fuzzes more.

To be clear the get will fail bc we arent putting in the key beforehand - so we check the error is not the "key not found" error and if so we have managed to break the smt with a fuzz. Hope this makes sense